pkg/http/routes.go

```diff
@@ -0,0 +1,32 @@
+package api
+
+import (
+	"fmt"
+	"log"
+	"net/http"
+)
+
+const (
+	// DefaultPort is the Default setting a port is set to for the server
+	// when not otherwise specified. It is publicly available for other
+	// clients to lookup, and potentially use.
+	DefaultPort = 8081
+)
+
+type StartOptions struct {
+	Port     int
+	BindHost string
+}
+
+// Start runs the cragcast API server.
+func Start(opts StartOptions) {
+	controller := New()
```

Comment:
is it worth renaming this function to `NewController` or `GenerateNewController`? `New` feels very ambiguous and it isn't obvious that it's coming from the `controllers.go` file unless you go to definition

Reply:
I agree this is rather unclear. I tried to not refactor this entirely, but here's how i'm thinking it should be:

We likely should move everything, for now, into a `server` package. `New()` should be on the server, which should then have a `Run()` method attached to it.

What are your thoughts on this?

Reply:
I think you meant to respond to https://github.com/dsauerbrun/cragcast/pull/5#discussion_r1178250882

Anyways, I think that's fine and makes sense if we're intent on having a cli entry point. IMO there isn't a real need for CLI entry as this is just a simple web server so instantiation of the server could just be part of a `StartServer` function in `cmd/main.go`. With all that being said, I'm sure there's a use case for some CLI commands so writing a server package is fine by me.

Reply:
that's fine; however,  my main gripe with this is that the function name is literally just `New`, new what?

In my mind, the controller is a close to the top layer for handling api endpoints. Eg. the `GetForecast` controller function will call some function like `GetAndProcessForecast("boulder-canyon")` and then decide what to respond with whether it be some type of error or the data.

There's gonna be a lot going on in that layer, mostly simple stuff(1to1 mapping of controller to endpoint) but still a lot so having a separate controller struct makes sense to me.


PS: I think we both might be confused and are talking about different things here.

Reply:
nah, this actually totally makes sense -- I'm going to switch it back to something more concrete as I agree `New()` is too ambiguous here.

Here's what my thinking was with `New` and why I think we should rethink our package structure/naming (at some point) to reintroduce it:

It's quite common to give a package a constructor, this is often, simply called `New`. From the [Package names section in Effective Go](https://go.dev/doc/effective_go#package-names):

> Similarly, the function to make new instances of ring.Ring—which is the definition of a constructor in Go—would normally be called NewRing, but since Ring is the only type exported by the package, and since the package is called ring, it's called just New, which clients of the package see as ring.New. Use the package structure to help you choose good names.

Thus, we should move into a model where there is a `server` package where `New` is created. Now whether we need handlers/controllers isolated in their own package inside `server` is TBD. Generally I'd advocate just keeping the package flat with separate files until we can concretely feel some pain that justifies that refactoring.

Reply:
Addressed via https://github.com/dsauerbrun/cragcast/pull/5/commits/cf917d25af5c9adc58a541cb514777393497d24c 👍

Reply:
`It's quite common to give a package a constructor, this is often, simply called New`

gotcha, makes sense now given that convention.

Agreed on keeping the package flat for now, no need to overengineer while everything is still simple.